### PR TITLE
822_form-fields-with-errors-a11y

### PR DESCRIPTION
**Pull request by jamesjacobs**
As raised in #822 screen readers didn't read form input help text/error messages.

The form input helpers have been updated so that when errors are passed to the helper the following happens:

- `aria-describedby="guid-<random-number>"` gets added to inputs.  `aria-describedby` supports a list of id's relating the elements that provide description so if multiple errors/help are present multiple `guid-`'s are listed.
- matching `id="guid-*"'s are added to any help blocks / errors.
- where an error is present `aria-invalid="true"` is added to the input 
- ~`role="alert"` is added to error help blocks~ Removed as they don't work if present on page load.

The code examples in the documentation have been updated.

Potentially contains a breaking change due to the new twig extension and changes in `app/app.php`.

**Will require docs publishing after release**

## Breaking change quick checklist

The following points should be considered as an early-warning of introducing a breaking change to a Continuum platform product. This is not an exhaustive list of what constitutes a breaking change.

| Does this pull request... | Yes / No |
| ---- | ---- |
| require changes to `main.js` | ✅ No |
| require changes to `index.js` | ✅ No |
| require changes to `pulsar.scss` | ✅ No |
| require changes to `gruntfile.js` | ✅ No |
| require changes to `package.json` (not including dev dependencies) | ✅ No |
| require changes to `base.html.twig` (or other core views like header/footer) | ✅ No |
| require changes to `app.php` (Twig loader) | 🛑 YES | 
| require new arguments to be passed to a js component | ✅ No |
| require markup changes to maintain functionality | ✅ No |

### Screenreaders

| Tool | Tested? | Known issues |
| ---- | ---- | ---- |
| JAWS | ✅ Yes |  |
| NVDA | ✅ Yes | Older NVDA versions <= 2017.1 only read one aria-describedby element |
| VoiceOver | ✅ Yes |  |

### Accessibility testing tools

| Tool | Tested? | Known issues |
| ---- | ---- | ---- |
| aXe | ✅ Yes |
| SiteImprove | ✅ Yes |
| Tenon.io | No |  |
| WAVE | ✅ Yes  |
| Lighthouse | ✅ Yes |

### CODE REVIEW

File: src/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtension.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
+<?php
+
+namespace Jadu\Pulsar\Twig\Extension;
+
+class HelperOptionsModifierExtension extends \Twig_Extension
+{
+	/**
```
Comment (Stanton) [inline src/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtension.php]: Mix of hard tabs and spaces in this file, can you make sure your editor is reading the .editorconfig file?

http://editorconfig.org/#download

File: src/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtension.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        if (array_key_exists('error', $options) and count($options['error']) > 0) {
+        	$numberOfErrors = count($options['error']);
+	    	$errorGuids = array();
+
+            // Set has_error for aria-invalid use
+            $options['has_error'] = true;
+
+            // Build guids for errors
+            for ($i = 0; $i < $numberOfErrors; $i++) { 
+                $errorGuids[] = 'guid-'.rand();
```
Comment (Stanton) [inline src/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtension.php]: spacing: `'guid-' .  rand();`

File: src/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtension.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            }
+
+            $options['error_ids'] = $errorGuids;
+        }
+
+        // Check if errors are present in options
+        if (array_key_exists('help', $options) and count($options['help']) > 0) {
+            $options['help_id'] = 'guid-'.rand();
```
Comment (Stanton) [inline src/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtension.php]: spacing

File: views/lexicon/forms/compound.html.twig [hunk 4]
```diff
@@ -301,15 +304,17 @@
             'id': 'comp-8',
             'help': 'For example: 25 12 1980',
             'label': 'Date of birth',
-            'error': ['Do this', 'Do this too'],
+            'error': ['Invalid day of birth entered', 'Invalid year of borth entered'],
+            'class': 'form__group--partial-state',
             'inputs': [
                 {
-                    'class': 'form__control-col--1',
+                    'class': 'form__control-col--1 has-error',
```
Comment (Stanton) [inline views/lexicon/forms/compound.html.twig]: We shouldn't need to explicitly add the error class if `has_error` is passed, the following example on L:341 doesn't...

File: views/pulsar/v2/helpers/elem.html.twig [hunk 5]
```diff
@@ -43,11 +43,29 @@ value       | string | Specifies the value of the input
...
 
+    {# Build aria-described string of guids for help blocks and errors #}
+    {% set describedByIds = [] %}
+
     {% if options.help_id is defined and options.help_id is not empty %}
-        {% set options = options|defaults({'aria-describedby': options.help_id}) %}
+        {% set describedByIds = describedByIds|merge([options.help_id])%}
+    {% endif %}
+
+    {% if options.error_ids is defined and options.error_ids is not empty %}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/elem.html.twig]: Is `error_ids` an internal-only value passed between macros or is it user-supplied? Can't see any reference to it

File: views/pulsar/v2/helpers/elem.html.twig [hunk 6]
```diff
@@ -43,11 +43,29 @@ value       | string | Specifies the value of the input
...
+        {% for id in options.error_ids %}
+            {% set describedByIds = describedByIds|merge([id])%}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/elem.html.twig]: Spacing `)%}`

File: views/pulsar/v2/helpers/elem.html.twig [hunk 7]
```diff
@@ -186,14 +204,31 @@ value       | string | Specifies the value of the input
...
+    {% set describedByIds = [] %}
 
     {% if options.help_id is defined and options.help_id is not empty %}
-        {% set options = options|defaults({'aria-describedby': options.help_id}) %}
+        {% set describedByIds = describedByIds|merge([options.help_id])%}
+    {% endif %}
+
+    {% if options.error_ids is defined and options.error_ids is not empty %}
+        {% for id in options.error_ids %}
+            {% set describedByIds = describedByIds|merge([id])%}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/elem.html.twig]: spacing `)%}`

File: views/pulsar/v2/helpers/elem.html.twig [hunk 8]
```diff
@@ -336,13 +371,31 @@ value       | string  | Specifies the value of the input
 {% spaceless %}
     {% import '@pulsar/pulsar/v2/helpers/util.html.twig' as util %}
 
+    {# Build aria-described string of guids for help blocks and errors #}
+    {% set describedByIds = [] %}
+
     {% if options.help_id is defined and options.help_id is not empty %}
-        {% set options = options|defaults({'aria-describedby': options.help_id}) %}
+        {% set describedByIds = describedByIds|merge([options.help_id])%}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/elem.html.twig]: spacing `)%}`

File: views/pulsar/v2/helpers/elem.html.twig [hunk 9]
```diff
@@ -336,13 +371,31 @@ value       | string  | Specifies the value of the input
...
+    {% endif %}
+
+    {% if options.error_ids is defined and options.error_ids is not empty %}
+        {% for id in options.error_ids %}
+            {% set describedByIds = describedByIds|merge([id])%}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/elem.html.twig]: spacing `)%}`

File: views/pulsar/v2/helpers/form.html.twig [hunk 10]
```diff
@@ -54,24 +54,53 @@ items         | array  | An array of items to put in the dropdown list (usually
...
-            }) 
-        %}
+    {% set options = modify_options(options|default({})) %}
+
+    {# Build aria-described string of guids for help blocks and errors #}
+    {% set describedByIds = [] %}
+
+    {% if options.error_ids is defined and options.error_ids is not empty %}
+        {% for id in options.error_ids %}
+            {% set describedByIds = describedByIds|merge([id])%}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/form.html.twig]: spacing `)%}`

File: views/pulsar/v2/helpers/form.html.twig [hunk 11]
```diff
@@ -54,24 +54,53 @@ items         | array  | An array of items to put in the dropdown list (usually
...
+        {% endfor %}
+    {% endif %}
+
+    {% if options.help_id is defined and options.help_id is not empty %}
+        {% set describedByIds = describedByIds|merge([options.help_id])%}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/form.html.twig]: spacing `)%}`

File: views/pulsar/v2/helpers/form.html.twig [hunk 12]
```diff
@@ -54,24 +54,53 @@ items         | array  | An array of items to put in the dropdown list (usually
...
+
+    {% if describedByIds|length %}
+        {% set options = options|defaults({'aria-describedby': describedByIds|join(' ')}) %}
+    {% endif %}
+
+    {# If error is preset add aria-invalid="true" and aria-describedby to all inputs #}
+    {% set items = options.items %}
+
+    {% if options.error is defined and options.error is not empty or options.help is defined and options.help is not empty %}
+        {% set newItems = [] %}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/form.html.twig]: minor tweak, but rename this to `tempItems` to hint that they're not exported

File: views/pulsar/v2/helpers/form.html.twig [hunk 13]
```diff
@@ -390,6 +417,21 @@ show-label | bool   | Control visibility of the `<label>` element without affect
...
+    {% set describedByIds = [] %}
+
+    {% if options.error_ids is defined and options.error_ids is not empty %}
+        {% for id in options.error_ids %}
+            {% set describedByIds = describedByIds|merge([id])%}
+        {% endfor %}
+    {% endif %}
+
+    {% if options.help_id is defined and options.help_id is not empty %}
+        {% set describedByIds = describedByIds|merge([options.help_id])%}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/form.html.twig]: spacing `)%}`

File: views/pulsar/v2/helpers/form.html.twig [hunk 14]
```diff
@@ -390,6 +417,21 @@ show-label | bool   | Control visibility of the `<label>` element without affect
...
     {% import '@pulsar/pulsar/v2/helpers/elem.html.twig' as elem %}
 
+    {% set options = modify_options(options|default({})) %}
+
+    {# Build aria-described string of guids for help blocks and errors #}
+    {% set describedByIds = [] %}
+
+    {% if options.error_ids is defined and options.error_ids is not empty %}
+        {% for id in options.error_ids %}
+            {% set describedByIds = describedByIds|merge([id])%}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/form.html.twig]: spacing `)%}`

File: views/pulsar/v2/helpers/form.html.twig [hunk 15]
```diff
@@ -551,20 +593,24 @@ https://jadu.github.io/pulsar/form-helpers/compound
...
-            }) 
-        %}
+    {% set options = modify_options(options|default({})) %}
+
+    {# Build aria-described string of guids for help blocks and errors #}
+    {% set describedByIds = [] %}
+
+    {% if options.error_ids is defined and options.error_ids is not empty %}
+        {% for id in options.error_ids %}
+            {% set describedByIds = describedByIds|merge([id])%}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/form.html.twig]: spacing

File: views/pulsar/v2/helpers/form.html.twig [hunk 16]
```diff
@@ -551,20 +593,24 @@ https://jadu.github.io/pulsar/form-helpers/compound
...
+        {% endfor %}
+    {% endif %}
+
+    {% if options.help_id is defined and options.help_id is not empty %}
+        {% set describedByIds = describedByIds|merge([options.help_id])%}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/form.html.twig]: spacing

File: views/pulsar/v2/helpers/form.html.twig [hunk 17]
```diff
@@ -926,24 +974,27 @@ value  | string, array  | The string, or array of strings to render inside the e
...
-        {% for error in errors %}
-            <span class="help-block is-error">
-                {{ html.icon('warning-sign') }} {{ error|raw -}}
-            </span>
-        {% endfor %}
-
+        {% if options.error_ids is defined and options.error_ids is not empty %}
+            {% for error in errors %}
+                <span class="help-block is-error" role="alert" id="{{ attribute(options.error_ids, loop.index0) }}">
+                    {{ html.icon('warning-sign') }} {{ error|raw -}}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/form.html.twig]: I know this is pre-existing but we should discuss whether we need `raw` here...

File: docs/_form-helpers/radio.md [hunk 18]
```diff
@@ -61,8 +61,8 @@ Any other options not listed here will be applied to the input.
     <div class="form__group has-error form-radio">
         <label for="inputRadioLeft" class="control__label">Default radio</label>
         <div class="controls">
-            <input id="inputRadioLeft" name="inputRadioLeft" checked="" type="radio" class="form__control radio">
-            <span class="help-block is-error"><i class="icon-warning-sign"></i> Something went wrong</span>
+            <input id="inputRadioLeft" name="inputRadioLeft" checked="" type="radio" aria-describedby="guid-12345678" class="form__control radio">
+            <span class="help-block is-error" id="guid-12345678" role="alert"><i class="icon-warning-sign"></i> Something went wrong</span>
```
Comment (Stanton) [inline docs/_form-helpers/radio.md]: SiteImprove may also want `aria-atomic` anywhere we use `role="alert"`

File: src/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtension.php [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+	 * Name of this extension
+	 *
+	 * @return string
+	 */
+	public function getName()
+    {
+        return 'modify_options';
+    }
+
+    public function getFunctions()
```
Comment (mikedevelops) [inline src/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtension.php]: Are we missing a docblock here?

File: src/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtension.php [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+	 * Returns a modified options array including error flag, error and help guids
+	 *
+	 * @param  array    $options    The form helper options
+	 * @return array
+	 */
+    public function modifyOptions($options)
+    {
+        // If we don't have $options, abandon ship
+        if ($options === null) {
+            return false;
```
Comment (mikedevelops) [inline src/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtension.php]: This return value should probably be specified in the docblock `@return array|boolean`

File: tests/unit/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtensionTest.php [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+
+    public function testNullReturnsFalse()
+    {
+        $this->assertEquals(false, $this->ext->modifyOptions(null));
+    }
+
+    public function testOptionsModifiedIfHelpSupplied()
+    {
+        $dataIn = array('help' => 'example help');
+        $dataOut = array('help' => 'example help', 'help_id' => 'guid-1');
```
Comment (mikedevelops) [inline tests/unit/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtensionTest.php]: does `$dataOut` need to be in any of these tests? Looks as though for each one it's defined and then unused. Were you planning on using it like ... ?
```
self::assertSame($dataOut['help'], $result['help']);
```

File: views/lexicon/forms/compound.html.twig (hunk 4 above)
Comment (jamesjacobs) [reply views/lexicon/forms/compound.html.twig]: Yes, `has_error` isn't connected to styling.

File: views/pulsar/v2/helpers/elem.html.twig (hunk 5 above)
Comment (jamesjacobs) [reply views/pulsar/v2/helpers/elem.html.twig]: It's internal-only. `error_ids` is created in the extension, which `group` calls to modify `options`, which then gets passed down to the macros in `elem.html.twig`.

File: views/lexicon/forms/compound.html.twig (hunk 4 above)
Comment (Stanton) [reply views/lexicon/forms/compound.html.twig]: I mean if the input uses `has_error: true`, then the `has-error` class should be applied automatically?

File: src/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtension.php (hunk 20 above)
Comment (jamesjacobs) [reply src/Jadu/Pulsar/Twig/Extension/HelperOptionsModifierExtension.php]: Added

File: docs/_form-helpers/radio.md (hunk 18 above)
Comment (jamesjacobs) [reply docs/_form-helpers/radio.md]: Added

File: docs/_form-helpers/text.md [hunk 22]
```diff
@@ -101,13 +101,13 @@ Use the `error` option when a form needs to highlight invalid or missing input,
     <div class="form__group has-error">
         <label for="inputTextError" class="control__label">First name</label>
         <div class="controls">
-            <input id="inputTextError" name="inputTextError" type="text" class="form__control">
-            <span class="help-block is-error"><i class="icon-warning-sign"></i> Please enter a name</span>
+            <input id="inputTextError" name="inputTextError" type="text" aria-describedby="guid-12345678" aria-invalid="true" class="form__control">
```
[the remark this answers is not in the document]
Comment (jamesjacobs) [reply docs/_form-helpers/text.md]: Pretty sure this is incorrect. The mark up validates and is read correctly by screen readers. WAVE, aXe, and SiteImprove don't report any issues with this.

[the remark this answers is not in the document]
Comment (jamesjacobs) [reply docs/_form-helpers/text.md]: Again, this looks like an accesslint error - you can see the label above is correctly referencing the input with `#inputTextError`

File: docs/_form-helpers/text.md [hunk 23]
```diff
@@ -121,13 +121,13 @@ The `has-success` can highlight fields that have been successfully updated.
     <div class="form__group has-success">
         <label for="inputTextSuccess" class="control__label">First name</label>
         <div class="controls">
-            <input id="inputTextSuccess" name="inputTextSuccess" type="text" class="form__control">
-            <span class="help-block">Optional help text</span>
+            <input id="inputTextSuccess" name="inputTextSuccess" type="text" aria-describedby="guid-12345678" class="form__control">
```
[the remark this answers is not in the document]
Comment (jamesjacobs) [reply docs/_form-helpers/text.md]: Pretty sure this is incorrect. The mark up validates and is read correctly by screen readers. WAVE, aXe, and SiteImprove don't report any issues with this.

File: docs/_form-helpers/text.md [hunk 24]
```diff
@@ -141,13 +141,13 @@ If a field value is changed by an interaction elsewhere in an interface, use the
     <div class="form__group has-changed">
         <label for="inputTextChanged" class="control__label">First name</label>
         <div class="controls">
-            <input id="inputTextChanged" name="inputTextChanged" type="text" class="form__control">
-            <span class="help-block">Optional help text</span>
+            <input id="inputTextChanged" name="inputTextChanged" type="text" aria-describedby="guid-12345678" class="form__control">
```
[the remark this answers is not in the document]
Comment (jamesjacobs) [reply docs/_form-helpers/text.md]: Pretty sure this is incorrect. The mark up validates and is read correctly by screen readers. WAVE, aXe, and SiteImprove don't report any issues with this.

File: docs/_form-helpers/text.md [hunk 25]
```diff
@@ -311,26 +311,26 @@ To maintain compliance with WCAG 2.0 AA, a form element must have a related labe
...
     </p>
 </div>
 
 <div class="pulsar-example form">
     <div class="form__group has-error">
         <label for="inputTextWithError" class="control__label">First name</label>
         <div class="controls">
-            <input id="inputTextWithError" type="text" class="form__control">
-            <span class="help-block is-error"><i aria-hidden="true" class="icon-warning-sign"></i> Please complete this field</span>
+            <input id="inputTextWithError" type="text" aria-describedby="guid-22334455" aria-invalid="true" class="form__control">
```
[the remark this answers is not in the document]
Comment (jamesjacobs) [reply docs/_form-helpers/text.md]: aria-describedby

File: docs/_form-helpers/text.md [hunk 26]
```diff
@@ -311,26 +311,26 @@ To maintain compliance with WCAG 2.0 AA, a form element must have a related labe
     <div class="form__group">
         <label for="inputTextWithHelp" class="control__label">First name</label>
         <div class="controls">
-            <input id="inputTextWithHelp" name="inputTextWithHelp" type="text" class="form__control">
-            <span class="help-block">You may enter more than one</span>
+            <input id="inputTextWithHelp" name="inputTextWithHelp" type="text" aria-describedby="guid-11223344" class="form__control">
```
[the remark this answers is not in the document]
Comment (jamesjacobs) [reply docs/_form-helpers/text.md]: Pretty sure this is incorrect. The mark up validates and is read correctly by screen readers. WAVE, aXe, and SiteImprove don't report any issues with this.

File: docs/_form-helpers/text.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (jamesjacobs) [reply docs/_form-helpers/text.md]: Pretty sure this is incorrect. The mark up validates and is read correctly by screen readers. WAVE, aXe, and SiteImprove don't report any issues with this.

File: docs/_form-helpers/text.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (jamesjacobs) [reply docs/_form-helpers/text.md]: Again, this looks like an accesslint error - you can see the label above is correctly referencing the input with #inputTextSuccess

File: docs/_form-helpers/text.md [hunk 27]
```diff
@@ -187,71 +187,71 @@ the width can be modified by passing the required column class via the `class` a
...
+            <input id="inputTextColOne" name="inputTextColOne" type="text" aria-describedby="guid-89012345" class="form__control">
+            <span class="help-block" id="guid-89012345">Optional help text</span>
         </div>
     </div>
     <div class="form__group form__control-col--9">
         <label for="inputTextColOne" class="control__label">First name</label>
         <div class="controls">
-            <input id="inputTextColOne" name="inputTextColOne" type="text" class="form__control">
-            <span class="help-block">Optional help text</span>
+            <input id="inputTextColOne" name="inputTextColOne" type="text" aria-describedby="guid-90123456" class="form__control">
```
[the remark this answers is not in the document]
Comment (jamesjacobs) [reply docs/_form-helpers/text.md]: Again, this looks like an accesslint error - you can see the label above is correctly referencing the input with #inputTextColOne

File: docs/_form-helpers/text.md (hunk 26 above)
[the remark this answers is not in the document]
Comment (jamesjacobs) [reply docs/_form-helpers/text.md]: Again, this looks like an accesslint error - you can see the label above is correctly referencing the input with #inputTextWithHelp

File: docs/_form-helpers/text.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (jamesjacobs) [reply docs/_form-helpers/text.md]: Again, this looks like an accesslint error - you can see the label above is correctly referencing the input with #inputTextWithError

**Pull request merged**